Feat/mutation testing

ASuciuX opened the pull request:
feature: mutation testing functional for PR differences in composite actions

wileyj commented on mutation-testing/check-packages-and-shards/action.yml:
```diff
@@ -0,0 +1,159 @@
+          !in_block
+        ' "$input_file" > "$temp_file"
+
+        # Check that the file exists before performing actions on it
+        if [ -s "$temp_file" ]; then
+          mv "$temp_file" "$input_file"
+        else
+          echo "Temp diff file missing!"
+          exit 1
+        fi
+
+        exit 0
```
small nit - this `else` is not needed. if the file exists, we can exit 0 or default to exit 1
```suggestion
        # Check that the file exists before performing actions on it
        if [ -s "$temp_file" ]; then
          mv "$temp_file" "$input_file"
          exit 0
        fi
        echo "Temp diff file missing!"
        exit 1
```

wileyj commented on mutation-testing/check-packages-and-shards/action.yml:
```diff
@@ -0,0 +1,159 @@
+      run: |
+        # Check that the file exists before performing actions on it
+        if [ ! -s git.diff ]; then
+          echo "Diff file missing!"
```
add in the file name here please, makes debugging easier

wileyj replied:
```suggestion
          echo "Diff file (git.diff) is missing or empty"
```

wileyj commented on mutation-testing/check-packages-and-shards/action.yml:
```diff
@@ -0,0 +1,159 @@
+
+        # Check that the file exists before performing actions on it
+        if [ ! -s all_mutants.txt ]; then
+          echo "The file containing mutants is missing or empty!"
```
```suggestion
          echo "The file containing mutants (all_mutants.txt) is missing or empty!"
```

wileyj commented on mutation-testing/check-packages-and-shards/action.yml:
```diff
@@ -0,0 +1,159 @@
+
+  steps:
+    - name: Checkout stacks-core repo
+      uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
```
for completeness, please add `id` keys to all steps

wileyj commented on mutation-testing/check-packages-and-shards/action.yml:
```diff
@@ -0,0 +1,159 @@
+      with:
+        fetch-depth: 0
+
+    - uses: taiki-e/install-action@ac89944b5b150d78567ab6c02badfbe48b0b55aa # v2.20.16
```
for completeness, please add `id` keys to all steps

wileyj commented on mutation-testing/check-packages-and-shards/action.yml:
```diff
@@ -0,0 +1,159 @@
+      with:
+        tool: cargo-mutants@23.12.2 # v23.12.2
+
+    - name: Relative diff
```
for completeness, please add `id` keys to all steps

wileyj commented on mutation-testing/check-packages-and-shards/action.yml:
```diff
@@ -0,0 +1,159 @@
+      run: |
+        git diff origin/${{ github.base_ref }}.. > git.diff
+
+    - name: Update git diff
```
for completeness, please add `id` keys to all steps

wileyj commented on mutation-testing/check-packages-and-shards/action.yml:
```diff
@@ -0,0 +1,159 @@
+
+        exit 0
+
+    - name: Split diffs
```
for completeness, please add `id` keys to all steps

wileyj commented on mutation-testing/output-pr-mutants/action.yml:
```diff
@@ -0,0 +1,166 @@
+  using: "composite"
+
+  steps:
+    - name: Download artifacts
```
for completeness, please add `id` keys to all steps

wileyj commented on mutation-testing/output-pr-mutants/action.yml:
```diff
@@ -0,0 +1,166 @@
+    - name: Download artifacts
+      uses: actions/download-artifact@v3
+
+    - name: Append output from shards
```
for completeness, please add `id` keys to all steps

wileyj commented on mutation-testing/output-pr-mutants/action.yml:
```diff
@@ -0,0 +1,166 @@
+
+        echo "$most_relevant_exit_code" > './mutants-shards/exit_code.txt'
+
+    - name: Print mutants
```
for completeness, please add `id` keys to all steps

wileyj commented on mutation-testing/pr-differences/action.yml:
```diff
@@ -0,0 +1,173 @@
+        ref: feat/mutation-testing
+        path: ./actions-repo
+
+    - name: Copy git diff
```
for completeness, please add `id` keys to all steps

wileyj commented on mutation-testing/pr-differences/action.yml:
```diff
@@ -0,0 +1,173 @@
+
+        mv git.diff ./actions-repo/mutation-testing/shell-scripts/
+
+    - uses: taiki-e/install-action@ac89944b5b150d78567ab6c02badfbe48b0b55aa # v2.20.16
```
for completeness, please add `id` keys to all steps

wileyj commented on mutation-testing/pr-differences/action.yml:
```diff
@@ -0,0 +1,173 @@
+      with:
+        tool: cargo-mutants@23.12.2 # v23.12.2
+
+    - name: Update git diff
```
for completeness, please add `id` keys to all steps

wileyj commented on mutation-testing/pr-differences/action.yml:
```diff
@@ -0,0 +1,173 @@
+      run: ./remove-deleted-file-lines.sh
+      working-directory: actions-repo/mutation-testing/shell-scripts
+
+    - name: Split diffs
```
for completeness, please add `id` keys to all steps

wileyj commented on mutation-testing/pr-differences/action.yml:
```diff
@@ -0,0 +1,173 @@
+        exit 0
+      working-directory: actions-repo/mutation-testing/shell-scripts
+
+    - name: Run mutants
```
for completeness, please add `id` keys to all steps

wileyj commented on mutation-testing/pr-differences/action.yml:
```diff
@@ -0,0 +1,173 @@
+        set -e
+      working-directory: actions-repo/mutation-testing/shell-scripts
+
+    - name: Upload artifact
```
for completeness, please add `id` keys to all steps

wileyj commented on mutation-testing/output-pr-mutants/action.yml:
```diff
@@ -0,0 +1,166 @@
+
+  steps:
+    - name: Download artifacts
+      uses: actions/download-artifact@v3
```
```suggestion
      uses: actions/download-artifact@6b208ae046db98c579e8a3aa621ab581ff575935 # v4.1.1
```

wileyj commented on mutation-testing/output-pr-mutants/action.yml:
```diff
@@ -0,0 +1,166 @@
+      shell: bash
+      run: |
+        # List of all possible mutant outcomes file names and folder names collected from jobs.
+        folders=("mutants-shard-big--1" "mutants-shard-big-0" "mutants-shard-big-1" "mutants-shard-big-2" "mutants-shard-big-3" "mutants-shard-big-4" "mutants-shard-big-5" "mutants-shard-big-6" "mutants-shard-big-7" "mutants-shard-small--1" "mutants-shard-small-0" "mutants-shard-small-1" "mutants-shard-small-2" "mutants-shard-small-3")
```
i think the first folder name here has an extra `-`. 

I'm also curious why these folder names are hardcoded - are these folders based on the the matrix of the calling workflow? if that is true, i think we can create a loop to do this step vs hardcoding the names. 


```suggestion
        folders=("mutants-shard-big-1" "mutants-shard-big-0" "mutants-shard-big-1" "mutants-shard-big-2" "mutants-shard-big-3" "mutants-shard-big-4" "mutants-shard-big-5" "mutants-shard-big-6" "mutants-shard-big-7" "mutants-shard-small--1" "mutants-shard-small-0" "mutants-shard-small-1" "mutants-shard-small-2" "mutants-shard-small-3")
```

wileyj replied:
ultimately, this is *fine* but i think it can/should be cleaner if possible.

ASuciuX replied:
'-1' is the default input for running without shards, hence the doubled dashes (https://github.com/stacks-network/actions/blob/feat/mutation-testing/mutation-testing/pr-differences/action.yml#L4-L7). 'mutants-shard-big-1' already exists and it's part of the matrix of shards.

The folders are hardcoded for the matrix jobs, how can i get the matrixes length in the composite action in order to use that directly in the 'for' loop?

wileyj replied:
you can do that via an input!
not a blocker for this PR, but perhaps a future improvement. 

say you have a calling workflow that uses a matrix of 8 jobs. 
when you call this composite workflow, you can send the number of jobs as an input:
```
uses: this workflow
with: 
  jobs: 8 
```
and then, in `output-pr-mutants/action.yml` you simple add an input with a sane default (8). 
in the shell, you can then do a simple for loop, i.e (pseudocode).:
```bash
for i in {1..${{inputs.matrix-jobs }}; do
  mkdir -p mutants-shard-big-$i
done
```

ref: https://github.com/orgs/community/discussions/26930#discussioncomment-3254002

i think i'm actually hardcoding this value myself in at least one place - so i'l have to test if this method works as i expect.

ASuciuX replied:
As we've discussed, I did it by parsing the type of operations it will do and that will know the number of shards it should have. This way, it will generate the specific folders, and only those will be tested in the workflows.

wileyj commented on mutation-testing/output-pr-mutants/action.yml:
```diff
@@ -0,0 +1,166 @@
+        # If the folder/file path exists, append the output to it's corresponding file name in a newly created folder that will contain all outputs combined
+        for file in "${files[@]}"; do
+          for folder in "${folders[@]}"; do
+            if [[ -s "$folder/$file" ]]; then
+              cat "$folder/$file" >> "mutants-shards/$file"
+            fi
```
```suggestion
            if [[ -s "$folder/$file" ]]; then
              cat "$folder/$file" >> "mutants-shards/$file"
            else
              echo "Missing file ($folder/$file)" 
            fi
```
not sure if this step should exit if a file is not found though

ASuciuX replied:
If i'm running without shards for each folder from 0 to 7 (big packages), respectively for 0 to 3 (small) i will get the 'Missing file' print 4 times for 12 folders. Only the artifacts with '-1' will be available and have outputs, so the print from the 'else' branch would be irrelevant. Also, if i'm running with shards, it will say 'Missing file' 4 times for each '-1' folder.

ASuciuX replied:
> not sure if this step should exit if a file is not found though

There will always be at least 2 missing folders due to the fact that it selects shards or no-shards from the complete list which includes both of them.

ASuciuX replied:
Just had an idea, i think we can parse from the main workflow the current case running through inputs ( if there are big packages, with/without shards, same for small ). Then, we know exactly which folders will be uploaded to artifacts and can make a dynamic list containing only the folders we need. That way we can also exit in the else branch since now we only check for folders we are certain should contain these mutant-outcomes files, and by not having them there we are sure something is wrong with the flow.

Referenced to this: https://github.com/stacks-network/actions/pull/8#discussion_r1449602978

wileyj commented on mutation-testing/output-pr-mutants/action.yml:
```diff
@@ -0,0 +1,166 @@
+              most_relevant_exit_code=$exit_code
+              ;;
+            esac
+          fi
```
what should happen if the file `$folder/exit_code.txt` does not exist?
should there be a default exit code maybe?

wileyj commented on mutation-testing/output-pr-mutants/action.yml:
```diff
@@ -0,0 +1,166 @@
+          fi
+        done
+
+        echo "$most_relevant_exit_code" > './mutants-shards/exit_code.txt'
```
please add a final exit code

wileyj commented on mutation-testing/output-pr-mutants/action.yml:
```diff
@@ -0,0 +1,166 @@
+        case $exit_code in
+          0)
+              if [[ -f ./mutants-shards/unviable.txt ]]; then
+                echo "Found unviable mutants!"
```
could this be sent to GITHUB_OUTPUT, or is the amount of data too large potentially?
this could be useful for debugging

ASuciuX replied:
Do you mean by `send to GITHUB_OUTPUT` to send the exit code, the message, or the mutants from the file? If you're talking about the exit code or log message, they are small and can be added. If it's about the mutant outcomes from the file, they are already printed to step summary.

wileyj commented on mutation-testing/output-pr-mutants/action.yml:
```diff
@@ -0,0 +1,166 @@
+            echo "All new and updated functions are caught!"
+            ;;
+          1)
+            echo "Invalid command line arguments!"
```
I would echo what args are being used for debugging

ASuciuX replied:
The args of the command are in another job, the 'pr-differences/action.yml' file, I don't have access to them in this job.

wileyj commented on mutation-testing/output-pr-mutants/action.yml:
```diff
@@ -0,0 +1,166 @@
+            exit 1
+            ;;
+          2 | 3)
+            echo "Found missed/timeout/unviable mutants!"
```
GITHUB_OUTPUT? or potentially too large?

ASuciuX replied:
Do you mean by `send to GITHUB_OUTPUT` to send the exit code, the message, or the mutants from the file? If you're talking about the exit code or log message, they are small and can be added. If it's about the mutant outcomes from the file, they are already printed to step summary.

wileyj commented on mutation-testing/output-pr-mutants/action.yml:
```diff
@@ -0,0 +1,166 @@
+          0)
+              if [[ -f ./mutants-shards/unviable.txt ]]; then
+                echo "Found unviable mutants!"
+                exit 1
```
```suggestion
                exit $exit_code
```

ASuciuX replied:
I will have to hardcode an exit code 5 here. In the case there are only unviable mutants, the `cargo mutants` command exits with code 0, which is the same as the case where all mutants are caught. That is why i am testing for unviable before proceeding to exit successfully.

wileyj replied:
strange, but understood - i've seen similar behaviour where errors/success both exit with 0 and it's annoying

wileyj commented on mutation-testing/output-pr-mutants/action.yml:
```diff
@@ -0,0 +1,166 @@
+            ;;
+          2 | 3)
+            echo "Found missed/timeout/unviable mutants!"
+            exit 1
```
```suggestion
                exit $exit_code
```

wileyj commented on mutation-testing/output-pr-mutants/action.yml:
```diff
@@ -0,0 +1,166 @@
+            ;;
+          4)
+            echo "Building the packages failed without any mutations!"
+            exit 1
```
```suggestion
                exit $exit_code
```

wileyj commented on mutation-testing/output-pr-mutants/action.yml:
```diff
@@ -0,0 +1,166 @@
+            ;;
+          *)
+            echo "Unknown exit code: $exit_code"
+            exit 1
```
```suggestion
                exit $exit_code
```

wileyj commented on mutation-testing/pr-differences/action.yml:
```diff
@@ -0,0 +1,173 @@
+      uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+      with:
+        repository: stacks-network/actions
+        ref: feat/mutation-testing
```
if we're merging this to `main`, be sure to update the ref here (not sure what the default is, but this key may not be needed if defatult is `main`)

wileyj commented on mutation-testing/pr-differences/action.yml:
```diff
@@ -0,0 +1,173 @@
+      shell: bash
+      run: |
+        if [ ! -s git.diff ]; then
+          echo "Diff file missing!"
```
if the file is not found/empty - we should exit 
```suggestion
          echo "Diff file missing!"
          exit 1
```

wileyj commented on mutation-testing/pr-differences/action.yml:
```diff
@@ -0,0 +1,173 @@
+    - uses: taiki-e/install-action@ac89944b5b150d78567ab6c02badfbe48b0b55aa # v2.20.16
+      name: Install cargo-mutants
+      with:
+        tool: cargo-mutants@23.12.2 # v23.12.2
```
is there a requirement for this version of `cargo-mutants`? if so, i would add an inline comment saying why this version is required

ASuciuX replied:
Using this specific version (currently the latest) because the commands written for it work as expected, and on a future release they may not work anymore or require different arguments/flags.

wileyj commented on mutation-testing/pr-differences/action.yml:
```diff
@@ -0,0 +1,173 @@
+      run: |
+        # Check that the file exists before performing actions on it
+        if [ ! -s git.diff ]; then
+          echo "Diff file missing!"
```
exit if file missing/empty
```suggestion
          echo "Diff file missing!"
          exit 1
```

wileyj commented on mutation-testing/pr-differences/action.yml:
```diff
@@ -0,0 +1,173 @@
+        fi
+
+        # Make a file containing all the mutants for the differences in the PR and a folder to split them into big and small packages
+        cargo mutants --in-diff git.diff --list > all_mutants.txt
```
ambivalent here, but i think you should add a return code check to verity the command ran successfully. 
2 ways to do it mainly:
1. 
```
if [ $? -ne 0 ]; then 
  echo "error"
  exit $?
fi
```
2. use a method like this, to wrap the return in a variable: https://github.com/actions-rust-lang/rustfmt/blob/main/action.yml#L23

wileyj commented on mutation-testing/pr-differences/action.yml:
```diff
@@ -0,0 +1,173 @@
+      working-directory: actions-repo/mutation-testing/shell-scripts
+
+    - name: Upload artifact
+      uses: actions/upload-artifact@v3
```
```suggestion
      uses: actions/upload-artifact@0b7f8abb1508181956e8e162db84b466c27e18ce # v3.1.2
```

wileyj commented on mutation-testing/pr-differences/action.yml:
```diff
@@ -0,0 +1,173 @@
+
+    - name: Upload artifact
+      uses: actions/upload-artifact@v3
+      if: always()
```
what would happen here if the path doesn't exist?
i'm not  a fan of using `always()`, have you tried alternatives like a conditional of previous steps?

ASuciuX replied:
Removed the `always()` - we don't need to upload an artifact if the previous steps failed, and we are also exiting early if something happens anyway.

wileyj replied:
Not seeing the change yet, but will resolve once it's commited

ASuciuX replied:
I’ve done the changes on a different branch and tested them there https://github.com/ASuciuX/stacks-core/pull/47. Did this in order to keep this cleaner and the commits on this as well. Please let me know if that’s alright.

wileyj commented on mutation-testing/check-packages-and-shards/README.md:
```diff
@@ -0,0 +1,28 @@
+# Check Packages and Shards action
+
+Checks whether to run mutants on big (`stackslib`/`stacks-node`) or small packages (all others), and whether to run them using strategy matrix or not.
```
can you add a link to the stacks-core repo for the stackslib and stacks-node packages?

wileyj commented on mutation-testing/output-pr-mutants/action.yml:
```diff
@@ -0,0 +1,166 @@
+name: Output Mutants
```
needs a description and optionally a badge

wileyj commented on mutation-testing/pr-differences/action.yml:
```diff
@@ -0,0 +1,173 @@
+name: PR Differences Mutants
```
needs a description and optionally a badge

wileyj commented:
one question i have here - is this workflow meant to run from *any* rust- based project, or only the `stacks-core` repo?
if it's only meant to run against stacks-core, i would advocate moving all the files under [stacks-core](https://github.com/stacks-network/actions/tree/feat/mutation-testing/stacks-core) (./stacks-core folder in the repo root). 

functionally it looks good - just minor changes to align with standards from other workflows in this repo

ASuciuX commented on mutation-testing/output-pr-mutants/action.yml:
```diff
@@ -0,0 +1,166 @@
+            fi
+
+            echo >> "$GITHUB_STEP_SUMMARY"
+          fi
```
We don't need an exit or an echo here because any file out of the 4 tested (missed, timeout, unviable, and caught) can miss, as long as there's at least 1 existing (in the case all are missing, the build fails earlier in the `Check Packages and Shards` job).

ASuciuX commented:
I've done the modifications on my repo (https://github.com/ASuciuX/actions/tree/test/move-shell-to-composite) in case there were more commits with fixes if the workflow fails. I've also removed the `shell-scripts` folder because the script was only called in one place, and with this, i've also removed the cloning of the `actions` repository in the composite.

> one question i have here - is this workflow meant to run from _any_ rust- based project, or only the `stacks-core` repo? if it's only meant to run against stacks-core, i would advocate moving all the files under [stacks-core](https://github.com/stacks-network/actions/tree/feat/mutation-testing/stacks-core) (./stacks-core folder in the repo root).
> 
> functionally it looks good - just minor changes to align with standards from other workflows in this repo

It was initially meant to only run on the `stacks-core` repo, but with the way it's structured right now, it can be run on any rust project. However, since only `stackslib` and `stacks-node` packages are considered `big`, it will run all the mutants as `small` - without `--test-threads 1` argument or `BITCOIND_TEST=1` and `RUST_BACKTRACE=full` env vars.

wileyj commented on mutation-testing/pr-differences/action.yml:
```diff
@@ -0,0 +1,173 @@
+      working-directory: actions-repo/mutation-testing/shell-scripts
+
+    - name: Upload artifact
+      uses: actions/upload-artifact@v3
```
```suggestion
      uses: actions/upload-artifact@1eb3cb2b3e0f29609092a73eb033bb759a334595 # v4.1.0
```

ASuciuX replied:
Did this change as well on the new PR https://github.com/ASuciuX/stacks-core/pull/47. It is more for testing and making sure everything works as expected without bloating to much the commits here.

wileyj reviewed (approved):
shipit
